File: android/src/com/android/tools/idea/res/AppResourceRepository.java
/*
 * Copyright (C) 2016 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.android.tools.idea.res;

import com.android.resources.aar.AarResourceRepository;
import com.android.tools.rendering.classloading.ModuleClassLoaderManager;
import com.android.tools.res.LocalResourceRepository;
import com.android.tools.res.ids.ResourceIdManager;
import com.google.common.annotations.VisibleForTesting;
import com.google.common.collect.ImmutableList;
import com.intellij.openapi.module.Module;
import com.intellij.openapi.util.Disposer;
import com.intellij.openapi.vfs.VirtualFile;
import java.util.Collection;
import java.util.List;
import org.jetbrains.android.facet.AndroidFacet;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.jetbrains.annotations.TestOnly;

/**
 * @see StudioResourceRepositoryManager#getAppResources()
 */
class AppResourceRepository extends MemoryTrackingMultiResourceRepository {
  private final AndroidFacet myFacet;
  private final Object RESOURCE_MAP_LOCK = new Object();

  /**
   * Resource directories. Computed lazily.
   */
  @Nullable private Collection<VirtualFile> myResourceDirs;

  @NotNull
  static AppResourceRepository create(@NotNull AndroidFacet facet, @NotNull Collection<AarResourceRepository> libraryRepositories,
                                      @NotNull LocalResourceRepository<VirtualFile> sampleDataResourceRepository) {
    AppResourceRepository repository =
        new AppResourceRepository(facet, computeLocalRepositories(facet, sampleDataResourceRepository), libraryRepositories);
    AndroidProjectRootListener.ensureSubscribed(facet.getModule().getProject());

    return repository;
  }

  @NotNull
  Collection<VirtualFile> getAllResourceDirs() {
    synchronized (RESOURCE_MAP_LOCK) {
      if (myResourceDirs == null) {
        ImmutableList.Builder<VirtualFile> result = ImmutableList.builder();
        for (LocalResourceRepository<VirtualFile> resourceRepository : getLocalResources()) {
          result.addAll(resourceRepository.getResourceDirs());
        }
        myResourceDirs = result.build();
      }
      return myResourceDirs;
    }
  }

  private static List<LocalResourceRepository<VirtualFile>> computeLocalRepositories(
      @NotNull AndroidFacet facet, @NotNull LocalResourceRepository<VirtualFile> sampleDataResourceRepository) {
    return ImmutableList.of(StudioResourceRepositoryManager.getProjectResources(facet), sampleDataResourceRepository);
  }

  private AppResourceRepository(@NotNull AndroidFacet facet,
                                @NotNull List<LocalResourceRepository<VirtualFile>> localResources,
                                @NotNull Collection<AarResourceRepository> libraryResources) {
    super(facet.getModule().getName() + " with modules and libraries");
    myFacet = facet;
    setChildren(localResources, libraryResources, ImmutableList.of(PredefinedSampleDataResourceRepository.getInstance()));
  }

  void updateRoots(@NotNull Collection<? extends AarResourceRepository> libraryResources,
                   @NotNull LocalResourceRepository<VirtualFile> sampleDataResourceRepository) {
    List<LocalResourceRepository<VirtualFile>> localResources = computeLocalRepositories(myFacet, sampleDataResourceRepository);
    updateRoots(localResources, libraryResources);
  }

  @VisibleForTesting
  void updateRoots(@NotNull List<? extends LocalResourceRepository<VirtualFile>> localResources,
                   @NotNull Collection<? extends AarResourceRepository> libraryResources) {
    synchronized (RESOURCE_MAP_LOCK) {
      myResourceDirs = null;
    }
    invalidateResourceDirs();
    setChildren(localResources, libraryResources, ImmutableList.of(PredefinedSampleDataResourceRepository.getInstance()));

    // Clear the fake R class cache and the ModuleClassLoader cache.
    Module module = myFacet.getModule();
    ResourceIdManager.get(module).resetDynamicIds();
    ResourceClassRegistry.get(module.getProject()).clearCache();
    ModuleClassLoaderManager.get().clearCache(module);
  }

  @TestOnly
  @NotNull
  static AppResourceRepository createForTest(@NotNull AndroidFacet facet,
                                             @NotNull List<LocalResourceRepository<VirtualFile>> modules,
                                             @NotNull Collection<AarResourceRepository> libraries) {
    AppResourceRepository repository = new AppResourceRepository(facet, modules, libraries);
    Disposer.register(facet, repository);
    return repository;
  }
}

File: android/src/com/android/tools/idea/res/AppResourceRepository.kt
/*
 * Copyright (C) 2016 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.android.tools.idea.res

import com.android.resources.aar.AarResourceRepository
import com.android.tools.idea.projectsystem.TestResourceResolutionToken
import com.android.tools.idea.projectsystem.getModuleSystem
import com.android.tools.idea.util.androidFacet
import com.android.tools.res.LocalResourceRepository
import com.intellij.openapi.Disposable
import com.intellij.openapi.vfs.VirtualFile
import org.jetbrains.android.facet.AndroidFacet
import org.jetbrains.android.uipreview.StudioModuleClassLoaderManager
import org.jetbrains.annotations.TestOnly
import org.jetbrains.annotations.VisibleForTesting

/** @see StudioResourceRepositoryManager.getAppResources */
@VisibleForTesting
class AppResourceRepository
private constructor(
  private val facet: AndroidFacet,
  parentDisposable: Disposable,
  localResources: List<LocalResourceRepository<VirtualFile>>? = null,
  libraryResources: Collection<AarResourceRepository>? = null,
) : MemoryTrackingMultiResourceRepository(parentDisposable, facet.module.name + " with modules and libraries") {
  private val resourceMapLock = Any()

  /** Resource directories. Computed lazily. */
  private var resourceDirs: List<VirtualFile>? = null

  val allResourceDirs: Collection<VirtualFile>
    get() {
      synchronized(resourceMapLock) {
        if (resourceDirs == null) {
          resourceDirs = localResources.flatMap { it.resourceDirs }
        }
        return requireNotNull(resourceDirs)
      }
    }

  init {
    val manager = StudioResourceRepositoryManager.getInstance(facet)
    setChildren(
      localResources ?: computeLocalRepositories(facet, manager),
      libraryResources ?: computeLibraryResources(manager),
      listOf(PredefinedSampleDataResourceRepository.getInstance()),
    )
  }

  override fun refreshChildren() {
    val manager = StudioResourceRepositoryManager.getInstance(facet)
    refreshChildren(computeLocalRepositories(facet, manager), computeLibraryResources(manager))
  }

  @VisibleForTesting
  fun refreshChildren(localResources: List<LocalResourceRepository<VirtualFile>>, libraryResources: Collection<AarResourceRepository>) {
    synchronized(resourceMapLock) { resourceDirs = null }
    invalidateResourceDirs()
    setChildren(localResources, libraryResources, listOf(PredefinedSampleDataResourceRepository.getInstance()))

    // Clear the fake R class cache and the ModuleClassLoader cache.
    val module = facet.module
    StudioResourceIdManager.get(module).resetDynamicIds()
    ResourceClassRegistry.get(module.project).clearCache()
    StudioModuleClassLoaderManager.get().clearCache(module)
  }

  companion object {
    @JvmStatic
    fun create(facet: AndroidFacet, parentDisposable: Disposable): AppResourceRepository {
      val repository = AppResourceRepository(facet, parentDisposable)
      AndroidProjectRootListener.ensureSubscribed(facet.module.project)

      return repository
    }

    private fun computeLocalRepositories(
      facet: AndroidFacet,
      manager: StudioResourceRepositoryManager,
    ): List<LocalResourceRepository<VirtualFile>> {
      val localRepositories = mutableListOf<LocalResourceRepository<VirtualFile>>()
      localRepositories.add(manager.projectResources)
      localRepositories.add(manager.sampleDataResources)

      // In screenshot testing, the test module is a separate module where the tests live
      // in its MAIN source set. For such modules to render correctly, we must include
      // the additional resource dependencies (like the production module) provided by the
      // project system.
      val androidModuleSystem = facet.module.getModuleSystem()
      val extraModules = TestResourceResolutionToken.getAdditionalLocalResourceDependencies(androidModuleSystem)
      for (extraModule in extraModules) {
        val extraFacet = extraModule.androidFacet
        if (extraFacet != null && extraFacet != facet) {
          localRepositories.add(StudioResourceRepositoryManager.getProjectResources(extraFacet))
        }
      }
      return localRepositories.distinct()
    }

    private fun computeLibraryResources(manager: StudioResourceRepositoryManager): Collection<AarResourceRepository> =
      manager.libraryResources

    @TestOnly
    @JvmStatic
    fun createForTest(
      facet: AndroidFacet,
      modules: List<LocalResourceRepository<VirtualFile>>,
      libraries: Collection<AarResourceRepository>,
    ) = AppResourceRepository(facet, parentDisposable = facet, modules, libraries)
  }
}
